include/session/network.hpp

```diff
@@ -0,0 +1,217 @@
+#pragma once
+
+#include <oxen/log/ring_buffer_sink.hpp>
+#include <oxen/quic.hpp>
+
+#include "network_service_node.hpp"
+#include "session/onionreq/builder.hpp"
+#include "session/onionreq/key_types.hpp"
+#include "session/types.hpp"
+
+namespace session::network {
+
+enum class ServiceNodeChangeType {
+    none = 0,
+    invalid_path = 1,
+    replace_swarm = 2,
+    update_path = 3,
+    update_node = 4,
+};
+
+struct onion_path {
+    std::shared_ptr<oxen::quic::connection_interface> conn;
+    std::vector<session::network::service_node> nodes;
+    uint8_t failure_count;
+};
+
+struct service_node_changes {
+    ServiceNodeChangeType type = ServiceNodeChangeType::none;
+    std::vector<session::network::service_node> nodes = {};
+    uint8_t path_failure_count = 0;
+    bool path_invalid = false;
+};
+
+struct request_info {
+    const service_node target;
+    const std::string endpoint;
+    const std::optional<ustring> body;
+    const std::optional<std::vector<service_node>> swarm;
+    const std::optional<onion_path> path;
+    const bool is_retry;
+};
+
+using network_response_callback_t = std::function<void(
+        bool success,
+        bool timeout,
+        int16_t status_code,
+        std::optional<std::string> response,
+        service_node_changes changes)>;
+
+class Network {
+  private:
+    oxen::quic::Network net;
+    std::shared_ptr<oxen::quic::GNUTLSCreds> creds;
+    std::vector<onion_path> paths;
+
+    std::shared_ptr<oxen::quic::Endpoint> endpoint;
+    std::shared_ptr<oxen::log::RingBufferSink> buffer;
+
+  public:
+    // Constructs a new network for the given credentials, all requests should be made via a single
+    // Network instance.
+    Network(const session::onionreq::ed25519_seckey ed25519_seckey);
```

Comment:
I've gone through and cleaned these all up - I have left a few `const` parameters in the C API functions just because Swift interprets it differently with/without the `const` (ie. we need a mutable pointer if it's not a `const` variable)